### ISSUE TITLE
sshfs GLib-CRITICAL error

### ISSUE THREAD

**Issue by dunn** [Homebrew/homebrew-fuse#16]
_From @knowledgecode on September 26, 2015 2:48_

I've been getting the following error message from `sshfs` since `glib` has been updated 2.46.0.
```sh
$ sshfs -V

(process:2435): GLib-CRITICAL **: g_slice_set_config: assertion 'sys_page_size == 0' failed
SSHFS version 2.5 (OSXFUSE SSHFS 2.5.0)
OSXFUSE library version: FUSE 2.7.3 / OSXFUSE 2.8.1
no mount point
```
It seems working fine for now, but the error message makes me worry. :(

_Copied from original issue: Homebrew/homebrew#44346_

**Comment by dunn** [comment]
Can you try building sshfs outside of Homebrew and see if it occurs there too?

**Comment by knowledgecode** [comment]
I tried to build the `sshfs-fuse-2.5.tar.gz`:
```sh
$ make
/Applications/Xcode.app/Contents/Developer/usr/bin/make  all-am
gcc -DHAVE_CONFIG_H -I.  -D_REENTRANT -DFUSE_USE_VERSION=26 -DLIBDIR=\"/usr/local/lib\"  -D_FILE_OFFSET_BITS=64 -D_DARWIN_USE_64_BIT_INODE -D_REENTRANT -I/usr/local/include/osxfuse/fuse -I/usr/local/Cellar/glib/2.46.0/include/glib-2.0 -I/usr/local/Cellar/glib/2.46.0/lib/glib-2.0/include -I/usr/local/opt/gettext/include  -g -O2 -Wall -W -MT sshfs-sshfs.o -MD -MP -MF .deps/sshfs-sshfs.Tpo -c -o sshfs-sshfs.o `test -f 'sshfs.c' || echo './'`sshfs.c
sshfs.c:1273:2: warning: 'sem_destroy' is deprecated [-Wdeprecated-declarations]
        sem_destroy(&req->ready);
        ^
/usr/include/sys/semaphore.h:53:5: note: 'sem_destroy' has been explicitly marked deprecated here
int sem_destroy(sem_t *) __deprecated;
    ^
sshfs.c:1840:2: warning: 'sem_init' is deprecated [-Wdeprecated-declarations]
        sem_init(&req->ready, 0, 0);
        ^
/usr/include/sys/semaphore.h:55:5: note: 'sem_init' has been explicitly marked deprecated here
int sem_init(sem_t *, int, unsigned int) __deprecated;
    ^
sshfs.c:3810:2: warning: 'g_thread_init' is deprecated [-Wdeprecated-declarations]
        g_thread_init(NULL);
        ^
/usr/local/Cellar/glib/2.46.0/include/glib-2.0/glib/deprecated/gthread.h:265:10: note: 'g_thread_init' has been explicitly marked deprecated here
void     g_thread_init                   (gpointer vtable);
         ^
3 warnings generated.
mv -f .deps/sshfs-sshfs.Tpo .deps/sshfs-sshfs.Po
gcc -DHAVE_CONFIG_H -I.  -D_REENTRANT -DFUSE_USE_VERSION=26 -DLIBDIR=\"/usr/local/lib\"  -D_FILE_OFFSET_BITS=64 -D_DARWIN_USE_64_BIT_INODE -D_REENTRANT -I/usr/local/include/osxfuse/fuse -I/usr/local/Cellar/glib/2.46.0/include/glib-2.0 -I/usr/local/Cellar/glib/2.46.0/lib/glib-2.0/include -I/usr/local/opt/gettext/include  -g -O2 -Wall -W -MT sshfs-cache.o -MD -MP -MF .deps/sshfs-cache.Tpo -c -o sshfs-cache.o `test -f 'cache.c' || echo './'`cache.c
mv -f .deps/sshfs-cache.Tpo .deps/sshfs-cache.Po
gcc -D_FILE_OFFSET_BITS=64 -D_DARWIN_USE_64_BIT_INODE -D_REENTRANT -I/usr/local/include/osxfuse/fuse -I/usr/local/Cellar/glib/2.46.0/include/glib-2.0 -I/usr/local/Cellar/glib/2.46.0/lib/glib-2.0/include -I/usr/local/opt/gettext/include  -g -O2 -Wall -W   -o sshfs sshfs-sshfs.o sshfs-cache.o  -L/usr/local/lib -L/usr/local/Cellar/glib/2.46.0/lib -L/usr/local/opt/gettext/lib -losxfuse -pthread -liconv -lgthread-2.0 -lglib-2.0 -lintl  
clang: warning: argument unused during compilation: '-pthread'

$ sudo make install
 ./install-sh -c -d '/usr/local/bin'
  /usr/bin/install -c sshfs '/usr/local/bin'
 ./install-sh -c -d '/usr/local/share/man/man1'
 /usr/bin/install -c -m 644 sshfs.1 '/usr/local/share/man/man1'...
```
But it didn't occur the error there.
```sh
$ sshfs -V
SSHFS version 2.5
OSXFUSE library version: FUSE 2.7.3 / OSXFUSE 2.8.1
no mount point
```

**Comment by jiyinyiyong** [comment]
Same problem here, what's happening?

**Comment by dltj** [comment]
For what its worth, it doesn't seem to be impacting the performance or functionality of sshfs...

**Comment by MinimaxHeadroom** [comment]
I'd also been receiving the same `glib` error anytime I used Homebrew-installed SSHFS. OSXFUSE is also installed: `brew cask install osxfuse`. Issuing `brew update && brew doctor` did nothing to fix the error.

I decided to uninstall both pieces of software, restart the system, and reinstall OSXFUSE (currently version 2.8.1) in the same manner as indicated above. I then rebooted again, as per Homebrew's install instructions.

Afterwards, I _manually downloaded and installed_ the [latest version of SSHFS](https://github.com/osxfuse/sshfs/releases/download/osxfuse-sshfs-2.5.0/sshfs-2.5.0.pkg) (2.5.0) from the [FUSE for OS X ](https://osxfuse.github.io/) site.

I'm now no longer receiving `glib` errors when using `sshfs` in Terminal.

> **Note**: I'm running OS X Yosemite v10.10.5 with the latest updates.

**Comment by wjaskowski** [comment]
I get the same error on OSX 10.9.5 when using sshfs from macports:
```
SSHFS version 2.5
OSXFUSE library version: FUSE 2.7.3 / OSXFUSE 2.7.5
```

**Comment by Sterbic** [comment]
Same error on happens on El Capitan (10.11.1).
````
sshfs -V
(process:1736): GLib-CRITICAL **: g_slice_set_config: assertion 'sys_page_size == 0' failed
SSHFS version 2.5 (OSXFUSE SSHFS 2.5.0)
OSXFUSE library version: FUSE 2.7.3 / OSXFUSE 2.8.2
no mount point
````

**Comment by MinimaxHeadroom** [comment]
@wjaskowski, @Sterbic, just in case, I recommend following [the steps I outlined earlier](https://github.com/Homebrew/homebrew-fuse/issues/16#issuecomment-149963414) to get an error-free setup. Still no idea why Homebrew- (MacPorts too, [apparently](https://github.com/Homebrew/homebrew-fuse/issues/16#issuecomment-155329164)) installed SSHFS is giving this error.

Also, today I upgraded OSXFUSE on Yosemite from v2.8.1 to v2.8.2 through System Preferences. SSHFS continues to work fine.

**Comment by DomT4** [comment]
> For whatever reason, I no longer receive the sshfs error in Terminal:

They'll have likely statically bundled everything into the binary, and the `glib` version it was built against was entirely compatible. The upstream `.pkg` binary is 5.7MB, compared to 76KB for the Homebrew version, which seems to confirm that.

**Comment by LuZhang-Lou** [comment]
@MinimaxHeadroom Your method worked, Thanks!

**Comment by bensowden** [comment]
@DomT4 is correct that they bundle all dependacies in the .pkg distro from osxfuse, it is stated on the [osxfuse wiki](https://github.com/osxfuse/osxfuse/wiki/SSHFS):
> The package installs a self-contained (as in "does not depend on external libraries") version of SSHFS 


It looks like the problem is that in the homebrew formula there is the line `depends_on "glib"` whereas it states in the README for [sshfs in osxfuse](https://github.com/osxfuse/sshfs) that:
> You also need to install the devel package for glib2.0.

which the `depends_on` line does not specify is required so when using hombrew we are compiling with the "stable 2.46.1" version of glib not the devel version sshfs specifically requires. My knowledge of homebrew is not good enough to know if this can be fixed. Does anyone know if this verison of glib is in homebrew and if it can be specifically required in thie formula?

**Comment by wenerme** [comment]
Same error

```
                 ###                  User: -
               ####                   Hostname: -
               ###                    Distro: OS X 10.11.2
       #######    #######             Kernel: Darwin
     ######################           Uptime: 34 days  1:03
    #####################             Shell: /usr/local/bin/bash
    ####################              Terminal: xterm-256color iTerm.app
    ####################              Packages: 167
    #####################             CPU: Intel Core i7-4980HQ CPU @ 2.80GHz
     ######################           Memory: 16 GB
      ####################            Disk: 59%
        ################              Battery: 100.00%
         ####     #####               IP Address: -
```

```
# sshfs -V

(process:52186): GLib-CRITICAL **: g_slice_set_config: assertion 'sys_page_size == 0' failed
SSHFS version 2.5 (OSXFUSE SSHFS 2.5.0)
OSXFUSE library version: FUSE 2.7.3 / OSXFUSE 2.8.2
no mount point
```

**Comment by danauerbach** [comment]
FWIW: This got rid of the error msg for me (and no longer relying on macports).

Started with:
OS X: v10.10.5
sshfs: v2.5.0 (via macports)
osxfuse: v2.8.1_0 (via macports)

sudo port uninstall sshfs
sudo port uninstall osxfuse

// Downloaded new osxfuse and sshfs install pkgs from https://osxfuse.github.io/
Install new osxfuse v2.8.3
reboot
install new sshfs v2.5.0
reboot

==> No longer receive error msg when mounted volume using sshfs.

**Comment by MinimaxHeadroom** [comment]
@danauerbach apart from the newer version numbers, how is this different than the steps I outlined [earlier](https://github.com/Homebrew/homebrew-fuse/issues/16#issuecomment-149963414)?

**Comment by danauerbach** [comment]
No different, just confirming it's effectiveness. Sorry, should have mentioned that in my post.

best,
dan

**Comment by MinimaxHeadroom** [comment]
No sweat, glad to hear you got it working.

**Comment by rspeed** [comment]
So is this an upstream issue in glib?

**Comment by wwilly** [comment]
No need to reboot between installing step of osxfuse and sshfs from github.

**Comment by hailthemelody** [comment]
Hi. I am having the same issue when installed via homebrew. Is the accepted solution to this to instead install via the PKG from https://osxfuse.github.io/ as previously mentioned? I am running 10.11.3 with the latest patches.
```
$ sshfs -V

(process:599): GLib-CRITICAL **: g_slice_set_config: assertion 'sys_page_size == 0' failed
SSHFS version 2.5 (OSXFUSE SSHFS 2.5.0)
OSXFUSE library version: FUSE 2.7.3 / OSXFUSE 2.8.3
no mount point
```

**Comment by vstakhov** [comment]
This issue is indeed related to glib. In fact, glib upstream marked `g_slice_set_config` deprecated and according to the code here: https://git.gnome.org/browse/glib/tree/glib/glib-init.c#n289 `glib_init` is executed when glib shared library is loaded, in turn `glib_init` indirectly initialize g_slice allocator but `g_slice_set_config` cannot be called after initialization. This means that `g_slice_set_config` is completely broken in glib and should not be ever used. The proposed patch: remove this function call completely (that should be safe in general).

**Comment by hailthemelody** [comment]
Thank you @vstakhov

**Comment by caryp** [comment]
Having same issue also.

**Comment by james-cole** [comment]
Following the advice of @MinimaxHeadroom I managed to fix this, just by unintalling the homebrew version of sshfs and installing the .dmg from the website.
Didn't need to uninstall oxsfuse or reboot.
Thanks!

**Comment by MikeMcQuaid** [comment]
Fixed in 6d3151cf2d93c42021385a6168acd96be3809cc4
